java/ql/src/Telemetry/AutomodelFrameworkModeCharacteristics.qll

```diff
@@ -0,0 +1,333 @@
+/**
+ * For internal use only.
+ */
+
+private import java
+private import semmle.code.Location as Location
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.TaintTracking
+private import semmle.code.java.security.PathCreation
+private import semmle.code.java.dataflow.ExternalFlow as ExternalFlow
+private import semmle.code.java.dataflow.internal.FlowSummaryImpl as FlowSummaryImpl
+private import semmle.code.java.security.ExternalAPIs as ExternalAPIs
+private import semmle.code.java.Expr as Expr
+private import semmle.code.java.security.QueryInjection
+private import semmle.code.java.security.RequestForgery
+private import semmle.code.java.dataflow.internal.ModelExclusions as ModelExclusions
+import AutomodelSharedCharacteristics as SharedCharacteristics
+import AutomodelEndpointTypes as AutomodelEndpointTypes
+
+/**
+ * A meta data extractor. Any Java extraction mode needs to implement exactly
+ * one instance of this class.
+ */
+abstract class MetadataExtractor extends string {
```

Comment:
Shouldn't this live in the shared logic library if the idea is to ensure that each mode will implement a metadata extractor with the the same desired signature?

Reply:
No, metadata is language specific. What the class extracts is Java-specific, eg., package, subtype.

When we integrate the Java application mode later, we should definitely try to use the same class for both (and all future) Java modes, though 👍

Reply:
Aren't all the `.ql/.qll` files living in `java/ql/src/...` Java specific? AIUI the `AutomodelFrameworkModeCharacteristics.qll` and `AutomodelApplicationModeCharacteristics.qll` we'll have there are both going to be for Java and using the `AutomodelSharedCharacterisrtics.qll` Java libraries.